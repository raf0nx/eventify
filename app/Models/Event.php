<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Event extends Model {
    use HasFactory;

    protected $fillable = [
        'name',
        'description',
        'image',
        'start_datetime',
    ];

    public function users(): \Illuminate\Database\Eloquent\Relations\BelongsToMany{
        return $this->belongsToMany(User::class, 'users_events');
    }
}
