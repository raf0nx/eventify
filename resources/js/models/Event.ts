import { User } from "./User";

export interface Event {
    created_at: Date;
    description: string;
    event_image: string;
    id: number;
    name: string;
    start_datetime: Date;
    updated_at: Date;
    users: User[];
}
