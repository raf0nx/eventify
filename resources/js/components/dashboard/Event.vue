<template>
	<v-lazy
		v-model="isActive"
		:options="{
			threshold: 0.5,
		}"
        min-height="200"
		transition="fade-transition"
	>
		<v-card class="my-12" elevation="8">
			<v-img
				lazy-src="https://picsum.photos/id/11/10/6"
				src="https://picsum.photos/id/11/500/300"
				height="300"
			></v-img>

			<v-card-subtitle class="pb-0">
				Event starts on {{ eventStartDate }} at {{ eventStartTime }}
			</v-card-subtitle>

			<v-card-title class="font-weight-bold">{{
				event.name
			}}</v-card-title>

			<v-card-text>
				<div>
					{{ event.description }}
				</div>

				<div class="pt-4">
					{{ event.users.length }} interested &bull; 0 going
				</div>
			</v-card-text>

			<v-card-actions class="py-4">
				<v-btn color="deep-purple" text>
					<v-icon left>mdi-star</v-icon>
					Interested
				</v-btn>
				<v-btn color="deep-purple" class="white--text font-weight-bold">
					<v-icon left>mdi-check-circle</v-icon>
					Going
				</v-btn>

				<v-spacer></v-spacer>

				<v-btn icon>
					<v-tooltip top>
						<template v-slot:activator="{ on, attrs }">
							<v-icon v-on="on" v-bind="attrs">mdi-heart</v-icon>
						</template>
						<span>Add to favourites</span>
					</v-tooltip>
				</v-btn>

				<v-btn icon>
					<v-tooltip top>
						<template v-slot:activator="{ on, attrs }">
							<v-icon v-on="on" v-bind="attrs">
								mdi-share-variant
							</v-icon>
						</template>
						<span>Share</span>
					</v-tooltip>
				</v-btn>
			</v-card-actions>
		</v-card>
	</v-lazy>
</template>

<script lang="ts">
	import { Vue, Component, Prop } from "vue-property-decorator";
	import moment from "moment";

	import { Event as EventModel } from "@/models/Event";

	@Component
	export default class Event extends Vue {
		@Prop()
		event!: EventModel;

        isActive = false;

		get eventStartDate(): string {
			return moment(this.event.start_datetime).format("dddd, MMM Do YYYY");
		}

		get eventStartTime(): string {
			return moment(this.event.start_datetime).format("h a");
		}
	}
</script>

<style>
</style>