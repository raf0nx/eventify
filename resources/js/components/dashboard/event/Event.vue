<template>
	<v-col class="col-xs-12 col-sm-6 col-lg-4 col-xl-3">
		<v-lazy
			v-model="isActive"
			:options="{
				threshold: 0.5,
			}"
			min-height="240"
			transition="fade-transition"
		>
			<v-hover v-slot="{ hover }">
				<v-card
					ripple
					class="rounded-lg"
					:elevation="hover ? 12 : 4"
					max-height="480"
					@click="goToEventDetails()"
				>
					<v-img
						lazy-src="https://picsum.photos/id/11/10/6"
						src="https://picsum.photos/id/11/500/300"
						max-height="200"
						class="rounded-t-lg"
					></v-img>

					<v-card-subtitle class="pb-0">
						{{ eventStartDate }} at {{ eventStartTime }}
					</v-card-subtitle>

					<v-card-title class="font-weight-bold">
						<router-link
							:to="{ name: 'Event', params: { id: event.id } }"
							class="
								event__link
								black--text
								text-decoration-none text-truncate
							"
							>{{ event.name }}</router-link
						>
					</v-card-title>

					<v-card-text>
						<v-tooltip bottom max-width="320">
							<template v-slot:activator="{ on, attrs }">
								<div
									class="text-truncate"
									v-on="on"
									v-bind="attrs"
								>
									{{ event.description }}
								</div>
							</template>
							<p class="text-justify mb-0">
								{{ event.description }}
							</p>
						</v-tooltip>
						<div class="pt-4">
							{{ event.users.length }} interested &bull; 0 going
						</div>
					</v-card-text>

					<v-card-actions class="pt-2 pb-4">
						<v-btn
							@click.stop="() => true"
							color="deep-purple"
							text
						>
							<v-icon left>mdi-star</v-icon>
							Interested
						</v-btn>
						<v-btn
							@click.stop="() => true"
							color="deep-purple"
							class="white--text font-weight-bold"
						>
							<v-icon left>mdi-check-circle</v-icon>
							Going
						</v-btn>

						<v-spacer></v-spacer>

						<v-btn icon>
							<v-tooltip top>
								<template v-slot:activator="{ on, attrs }">
									<v-icon
										@click.stop="() => true"
										color="red accent-2"
										v-on="on"
										v-bind="attrs"
										>mdi-heart-outline</v-icon
									>
								</template>
								<span>Add to favourites</span>
							</v-tooltip>
						</v-btn>
					</v-card-actions>
				</v-card>
			</v-hover>
		</v-lazy>
	</v-col>
</template>

<script lang="ts">
	import { Vue, Component, Prop } from "vue-property-decorator";
	import moment from "moment";

	import { Event as EventModel } from "@/models/Event";

	@Component
	export default class Event extends Vue {
		@Prop()
		event!: EventModel;

		isActive = false;

		get eventStartDate(): string {
			return moment(this.event.start_datetime).format("dddd, MMM Do YYYY");
		}

		get eventStartTime(): string {
			return moment(this.event.start_datetime).format("h a");
		}

		goToEventDetails(): void {
			setTimeout(() => {
				this.$router.push({
					name: "Event",
					params: { id: String(this.event.id) },
				});
			}, 200);
		}
	}
</script>

<style scoped>
	.event__link {
		border-bottom: 2px solid white;
		transition: border-bottom linear 0.5s;
	}
	.event__link:hover {
		border-bottom: 2px solid black;
	}
</style>