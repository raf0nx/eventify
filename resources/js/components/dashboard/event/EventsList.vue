<template>
	<v-col cols="9">
		<v-dialog
			v-model="showDialog"
			max-width="750px"
			transition="dialog-top-transition"
		>
			<event-stepper
				@closeDialog="showDialog = false"
				@rerenderEvents="getEvents()"
			></event-stepper>
		</v-dialog>
		<v-sheet
			class="mt-4 d-flex align-center rounded-lg justify-space-between"
			elevation="6"
			height="60"
		>
			<div
				class="headline grey--text text--darken-4 font-weight-bold ml-6"
			>
				Discover events
			</div>
			<v-tabs class="flex-grow-0 flex-shrink-0 mr-6" color="deep-purple">
				<v-tab class="font-weight-bold">Starting soon</v-tab>
				<v-tab class="font-weight-bold">Most popular</v-tab>
				<v-tab class="font-weight-bold">This week</v-tab>
			</v-tabs>
		</v-sheet>
		<action-chips @openDialog="showDialog = true" />
		<template v-if="events">
			<v-row>
				<event
					v-for="event in events"
					:event="event"
					:key="event.name"
				/>
			</v-row>
		</template>
		<template v-else>
			<v-row class="pr-4">
				<v-col
					v-for="i in 24"
					:key="i"
					class="col-xs-12 col-sm-6 col-lg-4 col-xl-3"
				>
					<v-skeleton-loader
						class="rounded-lg"
						type="image, text, article, table-cell, table-tfoot"
						:types="{
							article: 'heading, text',
							'table-tfoot': 'button@2, table-cell, avatar',
						}"
						max-height="480px"
						elevation="8"
					/>
				</v-col>
			</v-row>
		</template>
	</v-col>
</template>

<script lang="ts">
	import { Vue, Component } from "vue-property-decorator";

	import Event from "@/components/dashboard/event/Event.vue";
	import EventStepper from "@/components/dashboard/event/EventStepper.vue";
	import ActionChips from "@/components/dashboard/event/ActionChips.vue";
	import { Event as EventModel } from "@/models/Event";
	import EventService from "@/services/EventService";

	@Component({
		components: {
			Event,
			EventStepper,
			ActionChips,
		},
	})
	export default class EventsList extends Vue {
		events: EventModel[] | null = null;
		showDialog = false;

		async created(): Promise<void> {
			this.getEvents();
		}

		private async getEvents(): Promise<void> {
			const response = await EventService.getEvents();
			this.events = response.data;
		}
	}
</script>

<style>
	.dashboard .v-tabs {
		width: auto;
	}
	.dashboard .v-skeleton-loader__image {
		max-height: 200px !important;
	}
	.dashboard .v-skeleton-loader > .v-skeleton-loader__text {
		width: 75%;
		margin: 16px 16px 0;
	}
	.dashboard
		.v-skeleton-loader
		> .v-skeleton-loader__article
		> .v-skeleton-loader__heading {
		margin: 24px 0 24px 16px;
	}
	.dashboard
		.v-skeleton-loader
		> .v-skeleton-loader__article
		> .v-skeleton-loader__text {
		margin: 0 16px;
	}
	.dashboard .v-skeleton-loader > .v-skeleton-loader__table-cell {
		width: 124px;
		margin-left: 16px;
	}
	.dashboard .v-skeleton-loader > .v-skeleton-loader__table-tfoot {
		justify-content: unset;
	}
	.dashboard
		.v-skeleton-loader
		> .v-skeleton-loader__table-tfoot
		> .v-skeleton-loader__table-cell {
		flex: 1 1 auto;
		height: 0;
	}
	.dashboard
		.v-skeleton-loader
		> .v-skeleton-loader__table-tfoot
		.v-skeleton-loader__button:first-child {
		margin-left: 0;
	}
	.dashboard
		.v-skeleton-loader
		> .v-skeleton-loader__table-tfoot
		.v-skeleton-loader__button {
		width: 128px;
	}
	.dashboard
		.v-skeleton-loader
		> .v-skeleton-loader__table-tfoot
		> .v-skeleton-loader__avatar {
		flex: 1 0 auto;
	}
</style>