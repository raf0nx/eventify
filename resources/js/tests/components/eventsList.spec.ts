import { createLocalVue, shallowMount, Wrapper } from "@vue/test-utils";
import Vuetify from "vuetify";
import axios from "axios";

import EventsList from "@/components/dashboard/event/EventsList.vue";
import { Event } from "@/tests/constans/Event";

jest.mock("axios");

describe("EventsList.vue", () => {
    const localVue = createLocalVue();
    let wrapper: Wrapper<EventsList>;
    let vuetify: Vuetify;

    beforeEach(() => {
        vuetify = new Vuetify();
        // @ts-ignore
        axios.get.mockImplementationOnce(() =>
            Promise.resolve({ data: Event })
        );
        wrapper = shallowMount(EventsList, { localVue, vuetify });
    });

    afterEach(() => {
        jest.clearAllMocks();
        wrapper.destroy();
    });

    it("Should match snapshot", () => {
        // Assert
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("Should get events from API", () => {
        // Assert
        // @ts-ignore
        expect(wrapper.vm.events).toEqual(Event);
    });
});
