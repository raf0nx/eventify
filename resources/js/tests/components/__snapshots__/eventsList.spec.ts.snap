// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsList.vue Should match snapshot 1`] = `
"<v-col-stub cols=\\"9\\" tag=\\"div\\">
  <v-dialog-stub opendelay=\\"0\\" closedelay=\\"0\\" contentclass=\\"\\" maxwidth=\\"750px\\" origin=\\"center center\\" retainfocus=\\"true\\" transition=\\"dialog-top-transition\\" width=\\"auto\\">
    <event-stepper-stub></event-stepper-stub>
  </v-dialog-stub>
  <v-sheet-stub elevation=\\"6\\" height=\\"60\\" tag=\\"div\\" class=\\"mt-4 d-flex align-center rounded-lg justify-space-between\\">
    <div class=\\"headline grey--text text--darken-4 font-weight-bold ml-6\\">
      Discover events
    </div>
    <v-tabs-stub color=\\"deep-purple\\" activeclass=\\"\\" nexticon=\\"$next\\" previcon=\\"$prev\\" slidersize=\\"2\\" class=\\"flex-grow-0 flex-shrink-0 mr-6\\">
      <v-tab-stub ripple=\\"true\\" class=\\"font-weight-bold\\">Starting soon</v-tab-stub>
      <v-tab-stub ripple=\\"true\\" class=\\"font-weight-bold\\">Most popular</v-tab-stub>
      <v-tab-stub ripple=\\"true\\" class=\\"font-weight-bold\\">This week</v-tab-stub>
    </v-tabs-stub>
  </v-sheet-stub>
  <action-chips-stub></action-chips-stub>
  <v-row-stub tag=\\"div\\" class=\\"\\">
    <event-stub event=\\"Fri Jan 01 2021 01:00:00 GMT+0100 (Central European Standard Time)\\"></event-stub>
    <event-stub event=\\"Lorem ipsum dolor sit amet, consectetur adip\\"></event-stub>
    <event-stub event=\\"image.png\\"></event-stub>
    <event-stub event=\\"1\\"></event-stub>
    <event-stub event=\\"event\\"></event-stub>
    <event-stub event=\\"2021-01-01\\"></event-stub>
    <event-stub event=\\"Fri Jan 01 2021 01:00:00 GMT+0100 (Central European Standard Time)\\"></event-stub>
    <event-stub event=\\"1\\"></event-stub>
  </v-row-stub>
</v-col-stub>"
`;
